File: research/proto_benchmark/proto_benchmark/lib/generated_ffi/conversion.dart
// Generated conversions between proto models and FFI
// This layer is transparent to users

import 'dart:ffi';
import 'dart:typed_data';
import 'dart:convert';
import 'package:ffi/ffi.dart';

import 'proto.dart';
import 'ffi.dart';

extension UserConversions on User {
  /// Convert to FFI representation
  Pointer<UserFFI> toFFI() {
    final ptr = calloc<UserFFI>();
    final ref = ptr.ref;

    ref.user_id = this.user_id;
    ref.date_of_birth = this.date_of_birth;
    ref.created_at = this.created_at;
    ref.updated_at = this.updated_at;
    ref.account_balance = this.account_balance;
    ref.reputation_score = this.reputation_score;
    {
      final bytes = utf8.encode(this.username);
      final maxLen = 64;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.username[i] = bytes[i];
      }
      ref.username_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.email);
      final maxLen = 128;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.email[i] = bytes[i];
      }
      ref.email_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.first_name);
      final maxLen = 64;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.first_name[i] = bytes[i];
      }
      ref.first_name_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.last_name);
      final maxLen = 64;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.last_name[i] = bytes[i];
      }
      ref.last_name_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.display_name);
      final maxLen = 64;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.display_name[i] = bytes[i];
      }
      ref.display_name_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.bio);
      final maxLen = 512;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.bio[i] = bytes[i];
      }
      ref.bio_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.avatar_url);
      final maxLen = 256;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.avatar_url[i] = bytes[i];
      }
      ref.avatar_url_len = copyLen;
    }
    ref.is_verified = this.is_verified;
    ref.is_premium = this.is_premium;

    return ptr;
  }

  /// Free FFI pointer (call this when done)
  static void freeFFI(Pointer<UserFFI> ptr) {
    calloc.free(ptr);
  }
}

extension UserFromFFI on UserFFI {
  /// Convert from FFI representation
  static User fromFFI(Pointer<UserFFI> ptr) {
    final ref = ptr.ref;
    return User(
      user_id: ref.user_id,
      date_of_birth: ref.date_of_birth,
      created_at: ref.created_at,
      updated_at: ref.updated_at,
      account_balance: ref.account_balance,
      reputation_score: ref.reputation_score,
      username: () { final bytes = <int>[]; for (var i = 0; i < ref.username_len; i++) { bytes.add(ref.username[i]); } return utf8.decode(bytes); }(),
      email: () { final bytes = <int>[]; for (var i = 0; i < ref.email_len; i++) { bytes.add(ref.email[i]); } return utf8.decode(bytes); }(),
      first_name: () { final bytes = <int>[]; for (var i = 0; i < ref.first_name_len; i++) { bytes.add(ref.first_name[i]); } return utf8.decode(bytes); }(),
      last_name: () { final bytes = <int>[]; for (var i = 0; i < ref.last_name_len; i++) { bytes.add(ref.last_name[i]); } return utf8.decode(bytes); }(),
      display_name: () { final bytes = <int>[]; for (var i = 0; i < ref.display_name_len; i++) { bytes.add(ref.display_name[i]); } return utf8.decode(bytes); }(),
      bio: () { final bytes = <int>[]; for (var i = 0; i < ref.bio_len; i++) { bytes.add(ref.bio[i]); } return utf8.decode(bytes); }(),
      avatar_url: () { final bytes = <int>[]; for (var i = 0; i < ref.avatar_url_len; i++) { bytes.add(ref.avatar_url[i]); } return utf8.decode(bytes); }(),
      is_verified: ref.is_verified,
      is_premium: ref.is_premium,
    );
  }
}

extension PostConversions on Post {
  /// Convert to FFI representation
  Pointer<PostFFI> toFFI() {
    final ptr = calloc<PostFFI>();
    final ref = ptr.ref;

    ref.post_id = this.post_id;
    ref.user_id = this.user_id;
    ref.created_at = this.created_at;
    ref.updated_at = this.updated_at;
    ref.view_count = this.view_count;
    ref.like_count = this.like_count;
    ref.comment_count = this.comment_count;
    {
      final bytes = utf8.encode(this.username);
      final maxLen = 64;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.username[i] = bytes[i];
      }
      ref.username_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.title);
      final maxLen = 256;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.title[i] = bytes[i];
      }
      ref.title_len = copyLen;
    }
    {
      final bytes = utf8.encode(this.content);
      final maxLen = 4096;
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {
        ref.content[i] = bytes[i];
      }
      ref.content_len = copyLen;
    }
    ref.is_edited = this.is_edited;
    ref.is_pinned = this.is_pinned;

    return ptr;
  }

  /// Free FFI pointer (call this when done)
  static void freeFFI(Pointer<PostFFI> ptr) {
    calloc.free(ptr);
  }
}

extension PostFromFFI on PostFFI {
  /// Convert from FFI representation
  static Post fromFFI(Pointer<PostFFI> ptr) {
    final ref = ptr.ref;
    return Post(
      post_id: ref.post_id,
      user_id: ref.user_id,
      created_at: ref.created_at,
      updated_at: ref.updated_at,
      view_count: ref.view_count,
      like_count: ref.like_count,
      comment_count: ref.comment_count,
      username: () { final bytes = <int>[]; for (var i = 0; i < ref.username_len; i++) { bytes.add(ref.username[i]); } return utf8.decode(bytes); }(),
      title: () { final bytes = <int>[]; for (var i = 0; i < ref.title_len; i++) { bytes.add(ref.title[i]); } return utf8.decode(bytes); }(),
      content: () { final bytes = <int>[]; for (var i = 0; i < ref.content_len; i++) { bytes.add(ref.content[i]); } return utf8.decode(bytes); }(),
      is_edited: ref.is_edited,
      is_pinned: ref.is_pinned,
    );
  }
}

File: archive/proto2ffi-core-old/src/generator/conversion.rs
use crate::layout::{Layout, MessageLayout, FieldLayout};
use crate::error::Result;
use quote::quote;
use proc_macro2::TokenStream;
use std::path::Path;
use std::fs;

/// Generate conversion layer between proto models and FFI structs (Layer 3)
/// This makes FFI completely transparent to users
pub fn generate_rust_conversions(layout: &Layout, output: &Path) -> Result<()> {
    let mut tokens = TokenStream::new();

    for message in &layout.messages {
        tokens.extend(generate_message_conversions(message, layout));
    }

    let output_code = quote! {
        #![allow(dead_code)]

        use super::proto::*;  // Proto models
        use super::ffi::*;    // FFI structs

        #tokens
    };

    fs::write(output, output_code.to_string())?;

    // Format the output
    let _ = std::process::Command::new("rustfmt")
        .arg(output)
        .status();

    Ok(())
}

fn generate_message_conversions(message: &MessageLayout, layout: &Layout) -> TokenStream {
    let proto_name = syn::Ident::new(&message.name, proc_macro2::Span::call_site());
    let ffi_name = syn::Ident::new(&format!("{}FFI", message.name), proc_macro2::Span::call_site());

    // Generate to_ffi method
    let to_ffi_fields = message.fields.iter().map(|field| {
        generate_to_ffi_field(field, layout)
    });

    // Generate from_ffi method
    let from_ffi_fields = message.fields.iter().map(|field| {
        generate_from_ffi_field(field, layout)
    });

    quote! {
        impl #proto_name {
            /// Convert to FFI representation (zero-copy where possible)
            pub fn to_ffi(&self) -> #ffi_name {
                let mut ffi = #ffi_name::default();

                #(#to_ffi_fields)*

                ffi
            }

            /// Convert from FFI representation
            pub fn from_ffi(ffi: &#ffi_name) -> Self {
                Self {
                    #(#from_ffi_fields),*
                }
            }
        }
    }
}

fn generate_to_ffi_field(field: &FieldLayout, _layout: &Layout) -> TokenStream {
    let field_name = syn::Ident::new(&field.name, proc_macro2::Span::call_site());

    match field.rust_type.as_str() {
        // Primitives - direct copy
        "u8" | "u16" | "u32" | "u64" | "i8" | "i16" | "i32" | "i64" | "f32" | "f64" | "bool" => {
            quote! {
                ffi.#field_name = self.#field_name;
            }
        }

        // String field - copy to fixed array with length
        s if s.starts_with('[') && s.contains("u8") && !field.repeated => {
            // Extract array size from [u8; N]
            let len_field = syn::Ident::new(&format!("{}_len", field.name), proc_macro2::Span::call_site());

            quote! {
                {
                    let bytes = self.#field_name.as_bytes();
                    let copy_len = bytes.len().min(ffi.#field_name.len());
                    ffi.#field_name[..copy_len].copy_from_slice(&bytes[..copy_len]);
                    ffi.#len_field = copy_len as u32;
                }
            }
        }

        // Vec/repeated primitives - copy to fixed array with count
        s if !s.starts_with('[') && field.repeated => {
            let count_field = syn::Ident::new(&format!("{}_count", field.name), proc_macro2::Span::call_site());

            quote! {
                {
                    let copy_count = self.#field_name.len().min(ffi.#field_name.len());
                    ffi.#field_name[..copy_count].copy_from_slice(&self.#field_name[..copy_count]);
                    ffi.#count_field = copy_count as u32;
                }
            }
        }

        // Nested message - recursive conversion
        _ => {
            if field.repeated {
                let count_field = syn::Ident::new(&format!("{}_count", field.name), proc_macro2::Span::call_site());
                quote! {
                    {
                        let copy_count = self.#field_name.len().min(ffi.#field_name.len());
                        for i in 0..copy_count {
                            ffi.#field_name[i] = self.#field_name[i].to_ffi();
                        }
                        ffi.#count_field = copy_count as u32;
                    }
                }
            } else {
                quote! {
                    ffi.#field_name = self.#field_name.to_ffi();
                }
            }
        }
    }
}

fn generate_from_ffi_field(field: &FieldLayout, _layout: &Layout) -> TokenStream {
    let field_name = syn::Ident::new(&field.name, proc_macro2::Span::call_site());

    match field.rust_type.as_str() {
        // Primitives - direct copy
        "u8" | "u16" | "u32" | "u64" | "i8" | "i16" | "i32" | "i64" | "f32" | "f64" | "bool" => {
            quote! {
                #field_name: ffi.#field_name
            }
        }

        // String field - convert from fixed array using length
        s if s.starts_with('[') && s.contains("u8") && !field.repeated => {
            let len_field = syn::Ident::new(&format!("{}_len", field.name), proc_macro2::Span::call_site());

            quote! {
                #field_name: {
                    let len = ffi.#len_field as usize;
                    String::from_utf8_lossy(&ffi.#field_name[..len]).into_owned()
                }
            }
        }

        // Vec/repeated primitives - convert from fixed array using count
        s if !s.starts_with('[') && field.repeated => {
            let count_field = syn::Ident::new(&format!("{}_count", field.name), proc_macro2::Span::call_site());

            quote! {
                #field_name: {
                    let count = ffi.#count_field as usize;
                    ffi.#field_name[..count].to_vec()
                }
            }
        }

        // Nested message - recursive conversion
        _ => {
            let field_type = syn::Ident::new(&field.rust_type, proc_macro2::Span::call_site());

            if field.repeated {
                let count_field = syn::Ident::new(&format!("{}_count", field.name), proc_macro2::Span::call_site());
                quote! {
                    #field_name: {
                        let count = ffi.#count_field as usize;
                        (0..count)
                            .map(|i| #field_type::from_ffi(&ffi.#field_name[i]))
                            .collect()
                    }
                }
            } else {
                quote! {
                    #field_name: #field_type::from_ffi(&ffi.#field_name)
                }
            }
        }
    }
}

/// Generate Dart conversion extensions
pub fn generate_dart_conversions(layout: &Layout, output: &Path) -> Result<()> {
    let mut code = String::new();

    code.push_str("// Generated conversions between proto models and FFI\n");
    code.push_str("// This layer is transparent to users\n\n");
    code.push_str("import 'dart:ffi';\n");
    code.push_str("import 'dart:typed_data';\n");
    code.push_str("import 'dart:convert';\n");
    code.push_str("import 'package:ffi/ffi.dart';\n\n");
    code.push_str("import 'proto.dart';\n");
    code.push_str("import 'ffi.dart';\n\n");

    for message in &layout.messages {
        code.push_str(&generate_dart_message_conversions(message, layout));
        code.push_str("\n\n");
    }

    fs::write(output, code)?;
    Ok(())
}

fn generate_dart_message_conversions(message: &MessageLayout, _layout: &Layout) -> String {
    let class_name = &message.name;
    let ffi_class_name = format!("{}FFI", class_name);

    // toFFI method
    let to_ffi_fields = message.fields.iter().map(|field| {
        generate_dart_to_ffi_field(field)
    }).collect::<Vec<_>>().join("\n    ");

    // fromFFI factory
    let from_ffi_params = message.fields.iter().map(|field| {
        format!("      {}: {},", field.name, generate_dart_from_ffi_expr(field))
    }).collect::<Vec<_>>().join("\n");

    format!(
        r#"extension {}Conversions on {} {{
  /// Convert to FFI representation
  Pointer<{}> toFFI() {{
    final ptr = calloc<{}>();
    final ref = ptr.ref;

    {}

    return ptr;
  }}

  /// Free FFI pointer (call this when done)
  static void freeFFI(Pointer<{}> ptr) {{
    calloc.free(ptr);
  }}
}}

extension {}FromFFI on {} {{
  /// Convert from FFI representation
  static {} fromFFI(Pointer<{}> ptr) {{
    final ref = ptr.ref;
    return {}(
{}
    );
  }}
}}"#,
        class_name, class_name,
        ffi_class_name,
        ffi_class_name,
        to_ffi_fields,
        ffi_class_name,
        class_name, ffi_class_name,
        class_name, ffi_class_name,
        class_name,
        from_ffi_params
    )
}

fn extract_array_size(rust_type: &str) -> Option<usize> {
    // Extract size from "[T; N]" format
    if let Some(start) = rust_type.rfind(';') {
        if let Some(end) = rust_type.rfind(']') {
            let size_str = rust_type[start + 1..end].trim();
            return size_str.parse().ok();
        }
    }
    None
}

fn generate_dart_to_ffi_field(field: &FieldLayout) -> String {
    let field_name = &field.name;

    match field.dart_type.as_str() {
        "int" | "double" if !field.repeated => {
            format!("ref.{} = this.{};", field_name, field_name)
        }

        "String" if !field.repeated => {
            let array_size = extract_array_size(&field.rust_type).unwrap_or(256);
            format!(
                r#"{{
      final bytes = utf8.encode(this.{});
      final maxLen = {};
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {{
        ref.{}[i] = bytes[i];
      }}
      ref.{}_len = copyLen;
    }}"#,
                field_name, array_size, field_name, field_name
            )
        }

        "Uint8List" if !field.repeated => {
            let array_size = extract_array_size(&field.rust_type).unwrap_or(256);
            format!(
                r#"{{
      final bytes = this.{};
      final maxLen = {};
      final copyLen = bytes.length < maxLen ? bytes.length : maxLen;
      for (var i = 0; i < copyLen; i++) {{
        ref.{}[i] = bytes[i];
      }}
      ref.{}_len = copyLen;
    }}"#,
                field_name, array_size, field_name, field_name
            )
        }

        _ if field.repeated => {
            let array_size = field.max_count.unwrap_or(32);
            format!(
                r#"{{
      final maxCount = {};
      final copyCount = this.{}.length < maxCount ? this.{}.length : maxCount;
      for (var i = 0; i < copyCount; i++) {{
        ref.{}[i] = this.{}[i];
      }}
      ref.{}_count = copyCount;
    }}"#,
                array_size, field_name, field_name,
                field_name, field_name, field_name
            )
        }

        _ => {
            format!("ref.{} = this.{};", field_name, field_name)
        }
    }
}

fn generate_dart_from_ffi_expr(field: &FieldLayout) -> String {
    let field_name = &field.name;

    match field.dart_type.as_str() {
        "int" | "double" if !field.repeated => {
            format!("ref.{}", field_name)
        }

        "String" if !field.repeated => {
            format!(
                "() {{ final bytes = <int>[]; for (var i = 0; i < ref.{}_len; i++) {{ bytes.add(ref.{}[i]); }} return utf8.decode(bytes); }}()",
                field_name, field_name
            )
        }

        "Uint8List" if !field.repeated => {
            format!(
                "() {{ final bytes = <int>[]; for (var i = 0; i < ref.{}_len; i++) {{ bytes.add(ref.{}[i]); }} return Uint8List.fromList(bytes); }}()",
                field_name, field_name
            )
        }

        _ if field.repeated => {
            format!(
                "List.generate(ref.{}_count, (i) => ref.{}[i])",
                field_name, field_name
            )
        }

        _ => {
            format!("ref.{}", field_name)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::layout::{Layout, MessageLayout, FieldLayout};
    use std::collections::HashMap;

    #[test]
    fn test_generate_simple_conversion() {
        let layout = Layout {
            messages: vec![MessageLayout {
                name: "User".into(),
                size: 16,
                alignment: 8,
                fields: vec![
                    FieldLayout {
                        name: "id".into(),
                        rust_type: "u32".into(),
                        dart_type: "int".into(),
                        dart_annotation: "Uint32()".into(),
                        c_type: "uint32_t".into(),
                        offset: 0,
                        size: 4,
                        alignment: 4,
                        repeated: false,
                        max_count: None,
                    },
                    FieldLayout {
                        name: "name".into(),
                        rust_type: "[u8; 256]".into(),
                        dart_type: "Pointer<Utf8>".into(),
                        dart_annotation: "".into(),
                        c_type: "char".into(),
                        offset: 4,
                        size: 256,
                        alignment: 1,
                        repeated: false,
                        max_count: None,
                    },
                ],
                options: HashMap::new(),
            }],
            enums: vec![],
            alignment: 8,
        };

        let tokens = generate_message_conversions(&layout.messages[0], &layout);
        let code = tokens.to_string();

        assert!(code.contains("impl User"));
        assert!(code.contains("pub fn to_ffi"));
        assert!(code.contains("pub fn from_ffi"));
        assert!(code.contains("ffi . id = self . id"));
        assert!(code.contains("String :: from_utf8_lossy"));
    }

    #[test]
    fn test_dart_conversion_generation() {
        let message = MessageLayout {
            name: "User".into(),
            size: 16,
            alignment: 8,
            fields: vec![
                FieldLayout {
                    name: "id".into(),
                    rust_type: "u32".into(),
                    dart_type: "int".into(),
                    dart_annotation: "Uint32()".into(),
                    c_type: "uint32_t".into(),
                    offset: 0,
                    size: 4,
                    alignment: 4,
                    repeated: false,
                    max_count: None,
                },
            ],
            options: HashMap::new(),
        };

        let layout = Layout {
            messages: vec![message.clone()],
            enums: vec![],
            alignment: 8,
        };

        let code = generate_dart_message_conversions(&message, &layout);
        assert!(code.contains("extension UserConversions"));
        assert!(code.contains("toFFI()"));
        assert!(code.contains("fromFFI"));
    }
}
